### ISSUE TITLE
StreamSearcher buffers everything read?

### ISSUE THREAD

**Issue by ScottPeterJohnson** [libj/util#1]
Hey, neat library, and it also contains the only implementation of a character stream searcher I've been able to find in a library.

Unfortunately the implementation seems unnecessarily hobbled. Looking at its signature:
`public int search(final Reader in, final char[] buffer, final int offset) throws IOException {`
It requires a buffer to write everything it reads into, which means that in many cases you're still reading the entire input into memory! Plus, you have to presize the buffer beforehand- which is impossible in general for stream sources. 

Is there a reason it needs to rewrite into a buffer to perform a search?

**Comment by safris** [comment]
Hi @ScottPeterJohnson, thank you for this issue. You are 100% correct. I looked at this code, and it is indeed unnecessarily buffering everything that is read. It looks like this method was implemented for a specific purpose, to be used by some other application code that required the contents of the `Reader` to be buffered. Therefore, it's not really "hobbled", but rather "coupled". I've updated the implementation, and have added an overloaded `search(Reader)` method that foregoes the buffering. I hope this code is helpful for you!
